internal/engine/experiment/echcheck/measure.go

```diff
@@ -0,0 +1,114 @@
+package echcheck
+
+import (
+	"context"
+	"errors"
+	"github.com/ooni/probe-cli/v3/internal/measurexlite"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+	"net"
+	"net/url"
+	"time"
+)
+
+const (
+	testName      = "echcheck"
+	testVersion   = "0.1.0"
+	defaultDomain = "https://example.org"
+)
+
+var (
+	// errInputIsNotAnURL indicates that input is not an URL
+	errInputIsNotAnURL = errors.New("input is not an URL")
+
+	// errInvalidInputScheme indicates that the input scheme is invalid
+	errInvalidInputScheme = errors.New("input scheme must be https")
+)
+
+// TestKeys contains echcheck test keys.
+type TestKeys struct {
+	Control model.ArchivalTLSOrQUICHandshakeResult `json:"control"`
+	Target  model.ArchivalTLSOrQUICHandshakeResult `json:"target"`
+}
+
+// Measurer performs the measurement.
+type Measurer struct {
+	config Config
+}
+
+// ExperimentName implements ExperimentMeasurer.ExperiExperimentName.
+func (m *Measurer) ExperimentName() string {
+	return testName
+}
+
+// ExperimentVersion implements ExperimentMeasurer.ExperimentVersion.
+func (m *Measurer) ExperimentVersion() string {
+	return testVersion
+}
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	if measurement.Input == "" {
+		measurement.Input = defaultDomain
+	}
+	parsed, err := url.Parse(string(measurement.Input))
+	if err != nil {
+		return errInputIsNotAnURL
+	}
+	if parsed.Scheme != "https" {
+		return errInvalidInputScheme
+	}
+
+	// 1. perform a DNSLookup
+	trace := measurexlite.NewTrace(0, measurement.MeasurementStartTimeSaved)
+	resolver := trace.NewParallelDNSOverHTTPSResolver(sess.Logger(), m.config.resolverURL())
+	addrs, err := resolver.LookupHost(ctx, parsed.Host)
+	if err != nil {
+		return err
+	}
+	address := net.JoinHostPort(addrs[0], "443")
+
+	// 2. Set up TCP connections
+	var dialer net.Dialer
+	conn, err := dialer.DialContext(ctx, "tcp", address)
+	if err != nil {
+		return netxlite.NewErrWrapper(netxlite.ClassifyGenericError, netxlite.ConnectOperation, err)
+	}
+
+	conn2, err := dialer.DialContext(ctx, "tcp", address)
+	if err != nil {
+		return netxlite.NewErrWrapper(netxlite.ClassifyGenericError, netxlite.ConnectOperation, err)
+	}
+
+	// 3. Conduct and measure TLS Handshakes
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	testKeys := TestKeys{}
+	testKeys.Control = *handshake(ctx, conn, measurement.MeasurementStartTimeSaved, address, parsed.Host)
+	testKeys.Target = *handshakeWithEch(ctx, conn2, measurement.MeasurementStartTimeSaved, address, parsed.Host)
```

Comment:
Fixed in [30e6aba](https://github.com/ooni/probe-cli/pull/970/commits/30e6abac326bcff9713871198997e9e24c49a181).

Reply:
Thanks for changing the code so that the handshakes are executed in go routines!
Regarding the pattern you used: Did you think about data races in this case? IIUC, you are using the waitgroup `wg` to make sure that both handshakes finish before the result is read, is that correct?
I am asking because I think that it would probably be a bit cleaner to use channels instead but I want to make sure that I understand what you had in mind :)

Reply:
I do not see any potential for a data race here. The data shared between the go routines (start time, address and sni) is read-only and the return values are written into separate variables.

I am have trouble coming up with a cleaner way to do this using channels. Could you please elaborate on what you had in mind?

Reply:
I'll chime in for a second to comment on this specific point:

> I do not see any potential for a data race here. The data shared between the go routines (start time, address and sni) is read-only

Yes, because they're _read only_ there's no data race for the input variables.

> [...] and the return values are written into separate variables.

I am not convinced that the above argument is _sufficient_ to say you don't have a data race. IIUC, you also need to guarantee that reading the return variables, which you do on line ~109 when assembling `TestKeys` _happens after_ writing them in the goroutines. AFAICT, your code satisfies this _happens after_ property. Your `defer wg.Done()` _happens after_ writing the variables (by construction) and is guaranteed to _happen before_ `wg.Wait()` because of how the `sync.WaitGroup` primitive works for synchronizing goroutines. In turn, your initialization of them _happens after_ `wg.Wait()` (again by construction). I assume you omitted saying all of this for simplicity and you would agree with this analysis. However, in case I am wrong, I would like learn more about how you reasoned here.

> I am have trouble coming up with a cleaner way to do this using channels. Could you please elaborate on what you had in mind?

If the above analysis on the data race safety of your code is correct, your code is like this:

```Go
var a someType

wg.Add(1)
go func() {
  defer wg.Done()
  a = foo()
}

// HERE ....

wg.Wait()
b := a
```

As we said it's data-races free but there is an underlying problem. It's tricky to modify. If someone by mistake moves the assignment in the area marked as `HERE` you would now have a data race. I know this should not happen in theory, though there are many mistakes one could do with refactoring and this is still one of them. (We would probably catch this issue because we run tests with `-race` but that's beside the point.)

Compare the code above with this code:

```Go
ach := make(chan someType)

go func() {
  ach <- foo()
}

b := <-ach
```

This pattern has the same data-race-safe properties of the code above and _additionally_ it is impossible to refactor it in a way that the memory barrier provided by `wg.Wait` is decoupled from accessing the variable, because reading a channel is a memory barrier and a way to extract information from the goroutine _as part of the same operation_.

In your case, the code would probably look like this code:

```Go
ach1 := make(chan someType)
ach2 := make(chan someType)

go func() {
  ach1 <- foo()
}

go func() {
  ach2 <- bar()
}

b := <-ach1
c := <-ach2
```

Because you need anyway to wait for both goroutines to finish running, it does not matter much the relative order with which you read `b` and `c`. If `ach2` is written before `ach1`, you just wait for `ach1` to be written as well and then the read operation on the `ach2` channel would not block. Regardless, once you've read both channels, you have values of `b` and `c` that were obtained with a synchronization primitive that combines synchronization and message passing _and_ as such when reading the code you don't need to worry about the order of locking and reading.

Reply:
Thank you for the detailed explanation. I agree that the WaitGroup implementation is error prone as it depends on the order of blocking and reading. 

I have implemented the channel pattern you suggested in [c89d51c](https://github.com/ooni/probe-cli/pull/970/commits/c89d51c0502c49a3af94afa8982442d96c3154d1).